gunrock/app/mf/mf_test.cuh

```diff
@@ -7,324 +7,605 @@
-    #include <vector>
-    #include <utility>
+    #include <boost/graph/read_dimacs.hpp>
 #endif
 
+#include<queue>
+
 namespace gunrock {
 namespace app {
 namespace mf {
 
-/******************************************************************************
+/*****************************************************************************
  * Housekeeping Routines
- ******************************************************************************/
+ ****************************************************************************/
 
 /**
- * @brief Displays the SSSP result (i.e., distance from source)
- * @tparam T Type of values to display
- * @tparam SizeT Type of size counters
- * @param[in] preds Search depth from the source for each node.
- * @param[in] num_nodes Number of nodes in the graph.
+ * @brief Displays the MF result
+ *
+ * @tparam ValueT     Type of capacity/flow/excess
+ * @tparam VertxeT    Type of vertex 
+ * 
+ * @param[in] h_flow  Flow calculated on edges 
+ * @param[in] source  Index of source vertex
+ * @param[in] nodes   Number of nodes
  */
-template<typename T, typename SizeT>
-void DisplaySolution(T *array, SizeT length)
+template<typename GraphT, typename ValueT, typename VertexT>
+void DisplaySolution(GraphT graph, ValueT* h_flow, VertexT* reverse, 
+	VertexT sink, VertexT nodes)
 {
-    if (length > 40)
-        length = 40;
+    typedef typename GraphT::CsrT CsrT;
+    typedef typename GraphT::SizeT SizeT;
+    ValueT flow_incoming_sink = 0;
+    SizeT e_start = graph.CsrT::GetNeighborListOffset(sink);
+    SizeT num_neighbors = graph.CsrT::GetNeighborListLength(sink);
+    SizeT e_end = e_start + num_neighbors;
+    for (auto e = e_start; e < e_end; ++e)
+    {
+	ValueT flow = h_flow[reverse[e]];
+	if (util::isValid(flow))
+	    flow_incoming_sink += flow;
+    }
 
-    util::PrintMsg("[", true, false);
-    for (SizeT i = 0; i < length; ++i)
+    util::PrintMsg("The maximum amount of flow that is feasible to reach \
+	    from source to sink is " + std::to_string(flow_incoming_sink), 
+	    true, false);
+}
+/**
+ * @brief For given vertex v, find neighbor which the smallest height.  
+ *
+ * @tparam ValueT	Type of capacity/flow/excess
+ * @tparam VertxeT	Type of vertex
+ * @tparam GraphT	Type of graph
+ * @param[in] graph	Graph
+ * @param[in] x		Index of vertex 
+ * @param[in] height	Function of height on nodes 
+ * @param[in] capacity	Function of capacity on edges
+ * @param[in] flow	Function of flow on edges
+ *
+ * return Index the lowest neighbor of vertex x
+ */
+template<typename ValueT, typename VertexT, typename GraphT>
+VertexT find_lowest(GraphT graph, VertexT x, VertexT* height, ValueT* flow, 
+	VertexT source){
+    typedef typename GraphT::CsrT CsrT;
+    auto e_start = graph.CsrT::GetNeighborListOffset(x);
+    auto num_neighbors = graph.CsrT::GetNeighborListLength(x);
+    auto e_end = e_start + num_neighbors;
+    VertexT lowest;
+    VertexT lowest_id = -1; //graph.edges is unreachable value
```

Comment:
`VertexT lowest_id = (VertexT) -1;`
Or use `PreDefined` value and set it to invalid.